spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/Binding.java

```diff
@@ -31,141 +27,33 @@
-		this.endpoint = endpoint;
-		this.type = type;
-		this.properties = properties;
-	}
-
-	public static <T> Binding<T> forConsumer(String name, String group, AbstractEndpoint adapterFromBinder, T inputTarget,
-			DefaultBindingPropertiesAccessor properties) {
-		return new Binding<>(name, group, inputTarget, adapterFromBinder, Type.consumer, properties);
-	}
-
-	public static <T> Binding<T> forProducer(String name, T outputTarget, AbstractEndpoint adapterToBinder,
-			DefaultBindingPropertiesAccessor properties) {
-		return new Binding<>(name, null, outputTarget, adapterToBinder, Type.producer, properties);
-	}
-
-	public String getName() {
-		return name;
-	}
-
-	public String getGroup() {
-		return group;
-	}
-
-	public T getTarget() {
-		return target;
-	}
-
-	public AbstractEndpoint getEndpoint() {
-		return endpoint;
-	}
-
-	public Type getType() {
-		return type;
-	}
-
-	public DefaultBindingPropertiesAccessor getPropertiesAccessor() {
-		return properties;
-	}
-
-	@Override
-	public void start() {
-		endpoint.start();
-	}
-
-	@Override
-	public void stop() {
-		endpoint.stop();
-	}
-
-	@Override
-	public boolean isRunning() {
-		return endpoint.isRunning();
-	}
-
-	@Override
-	public String toString() {
-		return type + " Binding [name=" + name + ", target=" + target + ", endpoint=" + endpoint.getComponentName()
-				+ "]";
-	}
-
-	@Override
-	public int hashCode() {
-		final int prime = 31;
-		int result = 1;
-		result = prime * result + ((endpoint == null) ? 0 : endpoint.hashCode());
-		result = prime * result + ((group == null) ? 0 : group.hashCode());
-		result = prime * result + ((name == null) ? 0 : name.hashCode());
-		result = prime * result + ((properties == null) ? 0 : properties.hashCode());
-		result = prime * result + ((target == null) ? 0 : target.hashCode());
-		result = prime * result + ((type == null) ? 0 : type.hashCode());
-		return result;
-	}
-
-	@Override
-	public boolean equals(Object obj) {
-		if (this == obj)
-			return true;
-		if (obj == null)
-			return false;
-		if (getClass() != obj.getClass())
-			return false;
-		Binding<?> other = (Binding<?>) obj;
-		if (endpoint == null) {
-			if (other.endpoint != null)
-				return false;
-		} else if (!endpoint.equals(other.endpoint))
-			return false;
-		if (group == null) {
-			if (other.group != null)
-				return false;
-		} else if (!group.equals(other.group))
-			return false;
-		if (name == null) {
-			if (other.name != null)
-				return false;
-		} else if (!name.equals(other.name))
-			return false;
-		if (properties == null) {
-			if (other.properties != null)
-				return false;
-		} else if (!properties.equals(other.properties))
-			return false;
-		if (target == null) {
-			if (other.target != null)
-				return false;
-		} else if (!target.equals(other.target))
-			return false;
-		if (type != other.type)
-			return false;
-		return true;
-	}
-
+	/**
+	 * The name of the destination bound by the current binding.
+	 */
+	String getName();
```

Comment:
"I do think we need to avoid any reliance on the concrete DefaultBinding in those Binder impls as well though." I think it's the other way around, since the Binder impls create the bindings and will know exactly what kind of Binding they are working with. My idea here is to break the circular dependency by allowing Binders to subclass DefaultBinding and let it do its own cleanup code. Which basically means that the Binder will never see the Binding after creation anymore.  And viceversa.